Support RISC-V Compressed Instructions

xiaohan484 opened the pull request:
This work is based on sammer1107 and ccs100203. The RV32C is integrated with COMPUTE_GOTO and still don't support RV32C.F instruction. The C-extension compliance test was passed.

jserv commented:
You should rebase the latest `master` branch. See https://git-scm.com/docs/git-rebase
Also, decouple RISC-V Architecture Test from this pull request. That is, this pull request should only implement the revised RV32C instruction support.

jserv reviewed (changes requested):
Rework your git commits! See https://blog.yorkxin.org/posts/git-rebase.html

jserv commented:
Files statistics.[ch] should appear in another pull request!

eecheng87 commented:
@jserv , should these commits be squashed into a single one?

jserv commented:
> @jserv , should these commits be squashed into a single one?

It depends on the scenario. In this case, at least two git commits should be landed:
* C extension support
* Tweaks for computed-goto

In addition, @xiaohan484 should list @ccs100203 as the co-author. See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

eecheng87 commented:
@jserv, commits are squashed into 2 parts. @xiaohan484 will polish commit messages later.

eecheng87 commented:
@xiaohan484 , I think "co-author" should be mentioned in commit body, not commit header.

jserv commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "riscv-compliance"]
```
Don't do this. `riscv-compliance` should appear in another pull request.

xiaohan484 replied:
Sorry, this is my mistake when rebase the branch that appears, I will try to solve it.

jserv commented on Makefile:
```diff
@@ -54,10 +61,11 @@ OBJS = \
 	main.o \
 	syscall.o \
-	syscall_sdl.o
+	syscall_sdl.o 
```
Don't change this line. You should minimize the necessary changes.

xiaohan484 replied:
I think this is my mistake that type 'space' in this line. I will avoid these mistakes next time.

jserv commented on Makefile:
```diff
@@ -54,10 +61,11 @@ OBJS = \
 
 deps := $(OBJS:%.o=%.o.d)
 
+
```
Ditto. Minimize the necessary changes.

jserv commented on riscv.c:
```diff
@@ -773,67 +827,550 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+        else { /*imm==0 is reserved*/
+        }
+    } else {
+        // assert(!"Should be unreachbale.");
```
Why did you comment out `assert`?

xiaohan484 replied:
I think I leave a bad comment. That "assert" part is not unreachable. This part may be HINTS instruction. I leave it to "do nothing" in this part such as reserved instruction.

jserv commented on riscv.c:
```diff
@@ -773,67 +827,550 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+    const uint32_t rs1 = c_dec_rs1c(inst) | 0x08;
+
+    rv->X[rs1] &= imm;
+
```
Keep only one blank line.

jserv commented on riscv.c:
```diff
@@ -773,67 +827,550 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+static bool op_clwsp(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t temp = 0;
+    // temp |= ((inst & FCI_IMM_6_2) | 0b1110000) >>2;
```
Don't leave the dead code.

jserv commented on riscv.c:
```diff
@@ -773,67 +827,550 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+        assert(!"Should be unreachbale.");
+        break;
+    }
+
```
Ditto. Keep only one blank line.

jserv commented:
In git commit messages, you should append the statement "No RV32C.F support."

jserv reviewed (changes requested):
Read this article carefully: https://chiamakaikeanyi.dev/how-to-write-good-git-commit-messages/
Then, rewrite the git commit messages!

Replace `Co-authored-by: ccs100203` with `Co-authored-by: ccs100203 <ccs100203@gmail.com>`

jserv commented on riscv.c:
```diff
@@ -597,20 +647,22 @@ static bool op_system(struct riscv_t *rv, uint32_t inst)
         case 1:  // EBREAK
-            rv->io.on_ebreak(rv);
+
+            // rv->io.on_ebreak(rv);
```
Don't leave the dead code.

jserv commented on riscv.c:
```diff
@@ -863,25 +1417,37 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
         inst = rv->io.mem_ifetch(rv, rv->PC);
 
+        // Illegal instruction if inst[15:0] == 0
+        // assert(inst & 0xFFFF && "inst[15:0] must not be all 0.\n");
```
Would you rework the illegal instruction handling?

xiaohan484 replied:
If the illegal instruction occur, that will call rv_except_illegal_inst function:
1. record PC by csr_mepc
2. record instruction by csr_mtval
3. record except code by mscause
4. jump to base(csr_mtvec & ~0x3)

jserv replied:
You can add/improve comments around the illegal instruction handling in source files.

jserv commented on riscv_private.h:
```diff
@@ -210,3 +243,91 @@ static inline uint32_t sign_extend_b(uint32_t x)
 {
     return (int32_t)((int8_t) x);
 }
+// decode rs1 field
```
Add a new blank line before the comment.

jserv reviewed (changes requested):
Run `clang-format -i *.[ch]` and commit.

ccs100203 commented:
@xiaohan484, could you add `Uduru0522` into the co-author list in Integrate RV32C? He was another contributor for implementing compressed Instruction.

jserv commented on Makefile:
```diff
@@ -42,6 +43,12 @@ else
     VECHO = @printf
 endif
 
+ifeq ("$(DEBUG)","1")
```
Why did you add this? You should minimize the necessary changes.

jserv commented on README.md:
```diff
@@ -47,3 +47,4 @@ Add `-D` to enable and `-U` to disable the specific ISA extensions.
 ## License
 `rv32emu` is released under the MIT License.
 Use of this source code is governed by a MIT license that can be found in the LICENSE file.
+
```
Don't append new line.

jserv commented on elf.h:
```diff
@@ -47,8 +47,11 @@ bool elf_open(elf_t *e, const char *path);
 const char *elf_find_symbol(elf_t *e, uint32_t addr);
 
+/* get the range of .data section from the ELF file */
+bool elf_get_data_section_range(elf_t *e, uint32_t *start, uint32_t *end);
```
Why did you change elf.[ch]? Is this change related to RV32C?

jserv commented on main.c:
```diff
@@ -65,10 +65,7 @@ static void on_on_ecall(struct riscv_t *rv)
-{
-    rv_halt(rv);
-}
+static void on_on_ebreak(struct riscv_t *rv) {}
```
Avoid unrelated changes!

xiaohan484 replied:
The on_break function is not use after I solve the c.ebreak test.
- I treat the ebreak as the trap "break point"(in privilege specification) that pass the **c.ebreak** compliance test.
- Both(before/after modified) are **no pass** ebreak test in privilege mode.The I/M/Zifenci instruction test(sammer1107's compliance test passed) have **not** the "ebreak" test. 

Should I keep or remove related function?
- In the non-privilege mode the ebreak will let "the system back to the debug environment" But I don't know the detail in implementation
- The "on_ebreak" function let the system halt and terminate.(my understanding)

jserv replied:
The change against `on_on_ebreak` should appear in another pull request which addresses in the compliance and architecture test.

jserv commented on riscv.c:
```diff
@@ -773,60 +823,538 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
         /* dispatch this opcode */    \
         if (!op_##instr(rv, inst))    \
             goto quit;                \
+                                      \
```
Don't do this! Always minimize the necessary changes.

jserv commented on riscv.c:
```diff
@@ -773,60 +823,538 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+#define TABLE_TYPEC const c_opcode_t
 #endif
 
-    // clang-format off
```
Keep the format as the original did.

jserv commented on riscv.c:
```diff
@@ -773,60 +823,538 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+#define op_csw OP_UNIMP
+#endif  // ENABLE_RV32C
+
+
```
Leave only one blank line here.

jserv reviewed (changes requested):
We DO NEED informative git commit messages. Please check EACH git commit message at https://github.com/ccs100203/rv32emu-next/commits/shao .
You should minimize the necessary changes and squash the commits for reviewing along with the improved git commit message(s).

xiaohan484 commented:
> @xiaohan484, could you add `Uduru0522` into the co-author list in Integrate RV32C? He was another contributor for implementing compressed Instruction.

I added Uduru0522 into the co-author list, thank you for reminder.Please check the co-author list and let me know if there has any problem.

jserv reviewed (changes requested):
Move co-author list at the end of this git commit message.
Again, ensure that you have read the article: https://chiamakaikeanyi.dev/how-to-write-good-git-commit-messages/

The git commit message should mention:
1. Which instructions/parts you have implemented
2. The limitations and what this pull request lacks of
3. Implementation aspects such as computed-goto technique

jserv commented:
Drop the commit "Compliance Test and ComputeGoTo" since it should appear in another pull request.

jserv commented on riscv.c:
```diff
@@ -597,20 +644,20 @@ static bool op_system(struct riscv_t *rv, uint32_t inst)
             break;
         case 1:  // EBREAK
-            rv->io.on_ebreak(rv);
+            rv_breakpoint(rv, rv->PC);
```
You should explain why the change against `EBREAK` is associated with RV32C.

jserv replied:
The use of `rv->io.on_ebreak(rv)` was a matter of object-oriented programming in C as it encapsulates the unnecessary implementation details.

jserv commented on riscv.c:
```diff
@@ -773,53 +822,530 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
-        OP(store),  OP(store_fp), OP(unimp), OP(amo),      OP(op),     OP(lui),   OP(unimp), OP(unimp), // 01
-        OP(madd),   OP(msub),     OP(nmsub), OP(nmadd),    OP(fp),     OP(unimp), OP(unimp), OP(unimp), // 10
-        OP(branch), OP(jalr),     OP(unimp), OP(jal),      OP(system), OP(unimp), OP(unimp), OP(unimp), // 11
+        //  000         001           010        011           100         101
```
Keep the original indention since it improves the readability.

jserv commented on riscv.c:
```diff
@@ -863,25 +1406,34 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
         // fetch the next instruction
         inst = rv->io.mem_ifetch(rv, rv->PC);
 
+        //That will be checked in op If any illegal instruction occur
```
Improve the comment. Make it more informative!

jserv commented on elf.h:
```diff
@@ -47,7 +47,7 @@ bool elf_open(elf_t *e, const char *path);
 const struct Elf32_Sym *elf_get_symbol(elf_t *e, const char *name);
 
-/* Find symbole from a specified ELF file */
+/* Find symbol from a specified ELF file */
```
Don't change `elf.h` within this pull request. Instead, change in another pull request.

jserv commented on io.c:
```diff
@@ -70,7 +70,12 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
 {
     const uint32_t addr_lo = addr & mask_lo;
+
+#ifdef ENABLE_RV32C
```
You can add comment(s) which explains why such modification is necessary.

jserv commented:
File `README.md` should be modified to reflect RV32C. That is, "RISC-V RV32I[MAC]"

xiaohan484 commented:
> Move co-author list at the end of this git commit message. Again, ensure that you have read the article: https://chiamakaikeanyi.dev/how-to-write-good-git-commit-messages/
> 
> The git commit message should mention:
> 
> 1. Which instructions/parts you have implemented
> 2. The limitations and what this pull request lacks of
> 3. Implementation aspects such as computed-goto technique

Is this mean?Should I split the commit message to 2 or 3 commit?such like:

1. Add C-extension instruction length(16 bit instruction)support
- next PC can be 2 or 4

2. Integrate RV32C
- Compatible with uncompressed instruction
- RV32C.F doesn't support

3. Add Compute-goto support RV32C
- independent jump table for RV32C

jserv commented:
> Is this mean?Should I split the commit message to 2 or 3 commit?such like:

Exactly. You should split into several git commits and make each of them informative by writing comprehensive git commit messages.

ccs100203 committed:
Implement Preliminary RV32C support

  There has a variable inst_len(2 or 4) to calculate the next PC, RV32C
  related exception handler and all RV32C implementation exclude RV32C.F.

  Note: c.ebreak does'nt pass compliance test.

Co-authored-by: ccs100203 <ccs100203@gmail.com>
Co-authored-by: Uduru0522 <kurasiki.homura@gmail.com>

xiaohan484 committed:
Implement computed-goto path for RV32C

  This work use TABLE_TYPE_RVC type for RV32C's opcode handler type
  to dispatch function.All RV32C related function rename to op_c* to
  cooperate with computed-goto for standard uncompressed function.

xiaohan484 commented:
I try to split into several git message, and combine the message from https://github.com/ccs100203/rv32emu-next/commits/shao for reviewing.I think that might be more comprehensive and informative.

jserv commented on README.md:
```diff
@@ -35,6 +35,7 @@ should appear when Doom is loaded and executed.
 `rv32emu` is configurable, and you can modify `Makefile` to fit your expectations:
 * `ENABLE_RV32M`: Standard Extension for Integer Multiplication and Division
 * `ENABLE_RV32A`: Standard Extension for Atomic Instructions
+* `ENABLE_RV32C`: Standard Extension for Compressed Instructions(RV32C.F not support)
```
Change "(RV32C.F not support)" to "(RV32C.F excluded)"

jserv replied:
Add extra space character before `(`

jserv commented on riscv.c:
```diff
@@ -597,20 +642,20 @@ static bool op_system(struct riscv_t *rv, uint32_t inst)
             break;
         case 1:  // EBREAK
-            rv->io.on_ebreak(rv);
+            rv_breakpoint(rv,rv->PC);
```
Did you indent the source files with `clang-format -i`?

jserv replied:
By the way, you should explain the reason and consideration why you would like to perform such modification.

jserv commented on riscv.c:
```diff
@@ -770,15 +815,514 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+
+        if (imm != 0)
+            rv->X[rd] += imm;
+        else { /*imm==0 is reserved */
```
More reasons here.

jserv commented on riscv.c:
```diff
@@ -770,15 +815,514 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+        const int32_t imm = (tmp & 0x20000) ? (0xfffc0000 | tmp) : tmp;
+        if (imm != 0)
+            rv->X[rd] = imm;
+        else { /*imm==0 is reserved*/
```
Ditto.

jserv commented on riscv.c:
```diff
@@ -874,15 +1453,24 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
             // increment the cycles csr
             rv->csr_cycle++;
         } else {
             // TODO: compressed instruction
```
Remove "TODO" here since you provided the preliminary compressed instruction support.

jserv commented on riscv.c:
```diff
@@ -874,15 +1453,24 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
-            assert(!"Unreachable");
-        }
+            const uint16_t c_index = (inst & FR_C_15_13 >> 11) | (inst & FR_C_1_0);
+            // TODO: table implement
```
This comment looks ambiguous. Please improve it.

jserv commented on riscv_private.h:
```diff
@@ -210,3 +238,80 @@ static inline uint32_t sign_extend_b(uint32_t x)
+    temp |= (x & 0b0000000001100000) << 1;
+    temp |= (x & 0b0001000000000000) >> 4;
+    // extend to 16 bit
+    for(int i = 1; i <= 8; ++i){
```
Indent this file with `clang-format -i`.

xiaohan484 commented:
Because the ebreak/c.ebreak implementation is back to call rv->io.on_ebreak. I think I should remind that the compliance test "c.ebreak" will fail this time.

jserv commented on riscv.c:
```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+        case 3:  // AND
+            rv->X[rd] = rv->X[rs1] & rv->X[rs2];
+            break;
+        case 4:
```
Merge `case 4` and `case 5`.

jserv commented on riscv.c:
```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
 // opcode handler type
 typedef bool (*opcode_t)(struct riscv_t *rv, uint32_t inst);
+typedef bool (*c_opcode_t)(struct riscv_t *rv, uint16_t inst);
+
```
Keep only one blank line.

jserv commented on riscv_private.h:
```diff
@@ -104,6 +125,13 @@ struct riscv_t {
+        INST_UNKNOWN = 0,
+        INST_16 = 0x02,
+        INST_32 = 0x04,
+    }inst_len;
```
Indent all source files!

jserv replied:
Did you indent?

jserv commented on riscv.c:
```diff
@@ -803,23 +1304,39 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
     };
-// clang-format on
+    TABLE_TYPEC jump_tablec[] = {
+        //  00              01              10          11
```
Align the comment to ensure "00", "01", "10", and "11" starting with the corresponding `OP` positions.

jserv commented on riscv.c:
```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+    return true;
+}
+
+// static bool op_cXX(struct riscv_t *rv, uint16_t inst)
```
This comment is meaningless. Remove it.

jserv commented on riscv.c:
```diff
@@ -21,13 +21,34 @@ static void rv_except_inst_misaligned(struct riscv_t *rv, uint32_t old_pc)
     rv->csr_mcause = code;
 }
 
+static void rv_except_breakpoint(struct riscv_t *rv, uint32_t old_pc)
```
Is `rv_except_breakpoint` still making sense?

jserv commented on riscv.c:
```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+            rv->PC += rv->inst_len;
+            if (rd == rv_reg_zero)
+                rv->X[rv_reg_zero] = 0;
+        } else if (rs1 && rs2 == 0)  // rs1!=zero rs2== zero
```
Rewrite as following:
```c
} else if (rs1 && rs2 == 0) { // rs1 != zero AND rs2 == zero
```

jserv commented on riscv.c:
```diff
@@ -770,15 +815,469 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+                rv->X[rv_reg_zero] = 0;
+        } else if (rs1 && rs2 == 0)  // rs1!=zero rs2== zero
+        {
+            // C.JALR
```
The comment looks weird. Instead, you can quote from the specification.

jserv replied:
Should be `} else if (rs1 && rs2 == 0)  { // rs1 != zero AND rs2 == zero`

`{` appears before the comment.

jserv commented on riscv_private.h:
```diff
@@ -210,3 +238,80 @@ static inline uint32_t sign_extend_b(uint32_t x)
     return (int32_t)((int8_t) x);
 }
+
+// decode rs1 field
```
Use `#ifdef ENABLE_RV32C` and `#end` for conditional build of these compressed instruction specific helper functions.

jserv commented on riscv.c:
```diff
@@ -770,15 +794,465 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+    case 1:
+        if (rs1 == 0 && rs2 == 0)  // C.EBREAK
+            rv->io.on_ebreak(rv);
+        else if (rs1 && rs2)  // C.ADD
```
Ditto. Be aware of the position of `{` character.

jserv commented on riscv_private.h:
```diff
@@ -210,3 +238,92 @@ static inline uint32_t sign_extend_b(uint32_t x)
 }
+
+#ifdef ENABLE_RV32C
+
```
Remove this blank line.

jserv commented on riscv.c:
```diff
@@ -803,23 +1279,39 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
+        OP(csw),       OP(cbeqz),     OP(cswsp),  OP(unimp),  // 110
+        OP(cfsw),      OP(cbnez),     OP(cfswsp), OP(unimp),  // 111
+    };
+
```
Remove this blank line.

jserv commented on riscv.c:
```diff
@@ -770,15 +794,465 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+    const uint16_t rd = c_dec_rd(inst);
+    const uint32_t addr = rv->X[rv_reg_sp] + imm;
+
+    // reserve for rd==0
```
Should be "reserved"

jserv commented on riscv.c:
```diff
@@ -65,17 +65,34 @@ static void rv_except_store_misaligned(struct riscv_t *rv, uint32_t addr)
 {
-    /* TODO: dump more information */
-    assert(!"illegal instruction");
+    /* Jump to the base and record information to mcause */
```
Quote from RISC-V specification and describe Machine Cause Register (mcause).

jserv commented on riscv.c:
```diff
@@ -790,9 +1264,11 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
 #ifdef ENABLE_COMPUTED_GOTO
 #define OP(instr) &&op_##instr
 #define TABLE_TYPE const void *
+#define TABLE_TYPEC const void *
```
I am afraid that `TABLE_TYPEC` is ambiguous to reflect the use of compressed instructions. Can you think of another term?

xiaohan484 commented:
I rename TABLE_TYPEC to COMPRESSED_TABLE_TYPE and add more comment for exception handler function.Is it suitable?

jserv commented:
> I rename TABLE_TYPEC to COMPRESSED_TABLE_TYPE and add more comment for exception handler function.Is it suitable?

`COMPRESSED_TABLE_TYPE` is too long. How about `TABLE_TYPE_RVC`?

jserv commented on riscv.c:
```diff
@@ -25,19 +34,31 @@ static void rv_except_inst_misaligned(struct riscv_t *rv, uint32_t old_pc)
     rv->csr_mepc = rv->PC;
     rv->csr_mtval = addr;
 
+    // vector mode == 0,  DIRECT  : All exceptions set PC to base
```
Move the comment inlined. That is,
```c
switch (mode) {
case 0: // DIRECT: All exceptions set PC to base
    ...
case 1: VECTORED: Asynchronous interrupts set PC to BASE+4*
```

jserv commented on riscv.c:
```diff
@@ -25,19 +34,31 @@ static void rv_except_inst_misaligned(struct riscv_t *rv, uint32_t old_pc)
     rv->csr_mtval = addr;
 
+    // vector mode == 0,  DIRECT  : All exceptions set PC to base
+    // vector mode == 1,  VECTORED: Asynchonous interrupts set PC to BASE+4*
```
Typo. Should be "Asynchronous"

jserv commented on riscv.c:
```diff
@@ -47,19 +68,31 @@ static void rv_except_load_misaligned(struct riscv_t *rv, uint32_t addr)
         break;
     }
 
+    // mcause (Machine Cause Register)
```
Improve the writing!

jserv commented on riscv.c:
```diff
@@ -69,13 +102,44 @@ static void rv_except_store_misaligned(struct riscv_t *rv, uint32_t addr)
+        break;
+    }
+
+    // mcause (Machine Cause Register)
```
Ditto. Improve the writing.

jserv commented on riscv.c:
```diff
@@ -874,16 +1427,27 @@ void rv_step(struct riscv_t *rv, int32_t cycles)
+            const uint16_t c_index =
+                (inst & FR_C_15_13 >> 11) | (inst & FR_C_1_0);
+
+            // dispactch c_opcode (compress instruction)
```
It is "compressed instructions" rather than "compress instruction"

jserv commented on riscv_private.h:
```diff
@@ -210,3 +238,90 @@ static inline uint32_t sign_extend_b(uint32_t x)
+    CJ_IMM_3_1 = 0b0000000000111000,
+    CJ_IMM_5   = 0b0000000000000100,
+};
+// clang-format on
```
Why did you put `clang-format on` here?

jserv commented on riscv.c:
```diff
@@ -8,14 +8,23 @@
 
 static void rv_except_inst_misaligned(struct riscv_t *rv, uint32_t old_pc)
 {
+    // mtvec ( Machine Trap-Vector Base Address Register )
```
No extra space before `)` AND after `(`.

jserv commented on riscv.c:
```diff
@@ -770,15 +821,463 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+#define op_cfsd OP_UNIMP
 // opcode handler type
 typedef bool (*opcode_t)(struct riscv_t *rv, uint32_t inst);
+typedef bool (*c_opcode_t)(struct riscv_t *rv, uint16_t inst);
```
Add comments for RV32 and RVC typedef's accordingly.

jserv commented on riscv.c:
```diff
@@ -770,15 +821,463 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+            // C.JALR
+            // store the rv->X[rs1]
+            // make sure special case c.jalr ra
+            // jump to correct PC
```
The comment looks weird. Instead, you can quote from the specification and improve the writing.

jserv commented on riscv.c:
```diff
@@ -8,74 +8,118 @@
 
     switch (mode) {
-    case 0:  // DIRECT
+    case 0:  // DIRECT  : All exceptions set PC to base
```
should be "DIRECT: All exceptions set PC to base"

jserv commented on riscv.c:
```diff
@@ -8,74 +8,118 @@
+    rv->csr_mtval = inst;
+
+    switch (mode) {
+    case 0:  // DIRECT  : All exceptions set PC to base
```
Ditto.

jserv reviewed (changes requested):
Use "git rebase -i" to rework these commits.

xiaohan484 commented:
I rebase the commit but remain the latest commit.The code "inst_len" parts are in "struct riscv_t", I move "enum" part to the above of "struct riscv_t" and indent the enum parts.

jserv commented:
> I rebase the commit but remain the latest commit.The code "inst_len" parts are in "struct riscv_t", I move "enum" part to the above of "struct riscv_t" and indent the enum parts.

No, you should use `git rebase -i` to split and rework into two git commits:
1. Preliminary RV32C support
2. Implement computed-goto path for RV32C

Of course, you have to ensure each git commit message meaningful and informative.

jserv commented on riscv_private.h:
```diff
@@ -77,9 +77,36 @@ enum {
 
+enum {
+    INST_UNKNOWN = 0,
+    INST_16 = 0x02,
```
Use ordinary `2` and `4`.

jserv commented on riscv_private.h:
```diff
@@ -104,6 +131,9 @@ struct riscv_t {
     uint32_t csr_mbadaddr;
+    
+    // current instruction length
+    uint32_t inst_len;
```
Declare `inst_len` as `uint8_t` type since we are targeting RV32[IMAC].

jserv reviewed (changes requested):
Don't mention "RV32C.F excluded" in the subject of git commit messages.
You should write down the complete sentences rather than bullet points.

jserv commented:
Thank @xiaohan484 for contributing! Let's move forward to [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) compliance.
